### PR TITLE
Don't Suggest Labeling `const` and `unsafe` Blocks 

### LINKED ISSUE #128604: Trying to break from a `const` block gives invalid help message
### Code

```Rust
const _: () = const {
    break;
};
```

### Current output

```Shell
error[E0268]: `break` outside of a loop or labeled block
 --> src/lib.rs:2:5
  |
2 |     break;
  |     ^^^^^ cannot `break` outside of a loop or labeled block
  |
help: consider labeling this block to be able to break within it
  |
1 ~ const _: () = const 'block: {
2 ~     break 'block;
  |

For more information about this error, try `rustc --explain E0268`.
```

### Desired output

```Shell
Help suggestion should be removed.
```

### Rationale and extra context

The proposed syntax is invalid:
```Shell
error: expected expression, found keyword `const`
 --> src/lib.rs:1:15
  |
1 | const _: () = const 'block: {
  |               ^^^^^ expected expression
```

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.80.0 (051478957 2024-07-21)
binary: rustc
commit-hash: 051478957371ee0084a7c0913941d2a8c4757bb9
commit-date: 2024-07-21
host: x86_64-unknown-linux-gnu
release: 1.80.0
LLVM version: 18.1.7
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by veera-sivarajan**
Fixes #128604

Previously, both anonymous constant blocks (E.g. The labeled block
inside `['_'; 'block: { break 'block 1 + 2; }]`) and inline const
blocks (E.g. `const { ... }`) were considered to be the same
kind of blocks. This caused the compiler to incorrectly suggest
labeling both the blocks when only anonymous constant blocks can be
labeled.

This PR adds an other enum variant to `Context` so that both the
blocks can be handled appropriately.

Also, adds some doc comments and removes unnecessary `&mut` in a
couple of places.

### CODE REVIEW

File: tests/ui/inline-const/break-inside-inline-const-issue-128604.stderr [hunk 1]
```diff
@@ -0,0 +1,33 @@
+error[E0268]: `break` outside of a loop or labeled block
+  --> $DIR/break-inside-inline-const-issue-128604.rs:15:9
+   |
+LL |         break;
+   |         ^^^^^ cannot `break` outside of a loop or labeled block
```
Comment (estebank) [inline tests/ui/inline-const/break-inside-inline-const-issue-128604.stderr:5]: Was stderr file generated *after* the code changes were made? I'm just surprised that we seem to have the correct behavior here and no changes on the next commit.

Comment (veera-sivarajan) [reply tests/ui/inline-const/break-inside-inline-const-issue-128604.stderr:5]: Ah yes, I generated the tests after the changes. Will make sure to do it properly in the future.

**Comment by veera-sivarajan** [comment]
Just realized #128604 applies to `unsafe { break; }` as well.

I'll push some more changes in a bit to fix that in this PR itself.

**Pull request merged**